### DISCUSSION TITLE
3.57.3 trips on files saved form local URLs with mark-of-the-web ("production.ERROR: The specified HTML contains file://. This is not allowed")

### DISCUSSION THREAD

**Discussion by q--** [spatie/browsershot#683] [Bugs]
Hi, we're running into issues with the latest change that bans `file://` URLs ([this one](https://github.com/spatie/browsershot/commit/92cf16fc098211731f80d21687abeafbe2c457ad)). It appears that it works by looking for the string `file://` anywhere in the file.

We're using browsershot to render custom PDFs from various HTML templates. The problem is that templates for multiple users have the string `file://` somewhere in the HTML; in the file I examined, it came in the following format - a [Mark of the Web](https://superuser.com/questions/259636/after-saving-a-web-page-what-does-the-code-in-saved-from-url-mean) on the second line, after the document type

```html
<!-- saved from url=(0125)file:///C:/Users/PATH/FILENAME.html -->
```

(As an aside, I expect that it ended up there because at some point someone opened a local file in his browser (hence the `file://`-URL), and then saved it *from that browser* (hence the mark-of-the-web being set to said file URL).)

The error we're getting is as follows:

> production.ERROR: The specified HTML contains file://. This is not allowed. `{"userId":<redacted>,"exception":"[object] (Spatie\Browsershot\Exceptions\HtmlIsNotAllowedToContainFile(code: 0)`: The specified HTML contains file://. This is not allowed. at `/<redacted>/vendor/spatie/browsershot/src/Exceptions/HtmlIsNotAllowedToContainFile.php:11`)

----
Simple code to reproduce:
```php
$shot = Browsershot::html(<<< HTML_WITH_FILE_URL
<!doctype html>    
<!-- saved from url=(0125)file:///C:/Users/PATH/FILENAME.html -->
test
HTML_WITH_FILE_URL);
```

As an aside, this also occurs if the string `file://` occurs in the literal text of a document:

````php
$shot = Browsershot::html(<<< HTML_WITH_FILE_URL
<!doctype html>    
This file contains the text file://
HTML_WITH_FILE_URL);
````

**Comment by michaelklopf** [comment]
We have the same problem. Customers cannot use an important process now. Was introduced here 

https://github.com/spatie/browsershot/commit/cb106db5ff2b29e81dc8a61ffe0643991029c6f5

@freekmurze Any reasoning why this was added?

**Reply by freekmurze** [reply]
This has been changed to fix a security issue. Take a look at the relevant PR.

If this is urgent, feel free to PR a method to disable this security precaution. Make sure to include tests and updated readme.

**Reply by michaelklopf** [reply]
I'm sorry, I'm a little lost. I can't find a reference back to the PR. 

<img width="415" alt="image" src="https://user-images.githubusercontent.com/2892138/205902053-87b9be35-7934-4cd2-871e-622a401f9aad.png">
<img width="352" alt="image" src="https://user-images.githubusercontent.com/2892138/205902203-604bb6ee-fe98-426b-93db-baee2fe9cb71.png">

We need to find a solution for the problem first. I'm not sure if on customer side, or here in form of an option. We rolled back to 3.57.2 for now.

This is a breaking change though and I have the feeling that it needs different handling.

**Reply by q--** [reply]
For reference, we ended up working around this by doing a search/replace for 'file://' in the HTML before calling BrowserShot,
```php

        //Work around Browsershot throwing an error if it finds the string 'file://' anywhere in the HTML,
        //see https://github.com/spatie/browsershot/discussions/683
        $html = str_replace('file://', 'file_PROTOCOL_BLOCKED://', $html);
```

**Reply by michaelklopf** [reply]
We did that too

```
Str::remove(['file://', 'File://'], $matches[2]);
```

**Reply by epixian** [reply]
I was able to bypass the check using `Browsershot::htmlFromFilePath($path)`...

**Comment by cyreb7** [comment]
I just ran into this too, a user decided to put `file:///` into a text field. It would be nice if the library was better able to only check file links instead of blindly disallowing it through the whole document.

Could we use [Puppeteer Request Interception](https://pptr.dev/guides/network-interception) to block file requests instead?
